File: main/4sum-ii/4sum-ii-fast.cpp
class Solution {
public:
    int fourSumCount(vector<int>& __restrict a, vector<int>& __restrict b,
                     vector<int>& __restrict c, vector<int>& __restrict d)
    {
        unordered_map<int, int> ab_counts;
        for (const auto w : a)
            for (const auto x : b) ++ab_counts[w + x];

        const auto not_found = end(ab_counts);

        auto count = 0;
        for (const auto y : c) {
            for (const auto z : d) {
                const auto it = ab_counts.find(-(y + z));
                if (it != not_found) count += it->second;
            }
        }

        return count;
    }
};

File: main/4sum-ii/4sum-ii-fast.java
class Solution {
    private static final Integer ZERO = Integer.valueOf(0);

    public int fourSumCount(final int[] a, final int[] b,
                            final int[] c, final int[] d)
    {
        final Map<Integer, Integer> ab_sums = new HashMap<>();
        for (final int w : a) {
            for (final int x : b) {
                final Integer sum = w + x;
                ab_sums.put(sum, ab_sums.getOrDefault(sum, ZERO) + 1);
            }
        }

        int count = 0;
        for (final int x : c) {
            for (final int y : d)
                count += ab_sums.getOrDefault(-(x + y), ZERO);
        }

        return count;
    }
}
